src/joypad.c

```diff
@@ -0,0 +1,1217 @@
+
+    joybus_exec(input, output);
+    const joybus_cmd_n64_controller_read_port_t *recv_cmd =
+        (void *)&output[port];
+
+    return joypad_inputs_from_n64_controller_read(recv_cmd);
+}
+
+/**
+ * @brief Initialize the Joypad subsystem.
+ * 
+ * Starts reading Joypads during VI interrupt.
+ */
+void joypad_init(void)
+{
+    // Just increment the refcount if already initialized
+	if (joypad_init_refcount++ > 0) { return; }
+
+    // Initialize the timer subsystem (or increment its refcount)
+    timer_init();
+
+    JOYPAD_PORT_FOREACH (port)
+    {
+        joypad_device_reset(port, JOYBUS_IDENTIFIER_UNKNOWN);
+    }
+    register_VI_handler(joypad_vi_interrupt_callback);
+    joypad_identify_sync(true);
+    joypad_read_sync();
+}
+
+/**
+ * @brief Close the Joypad subsystem.
+ * 
+ * Stops reading Joypads during VI interrupt.
+ */
+void joypad_close(void)
+{
+    // Do nothing if there are still dangling references.
+	if (--joypad_init_refcount > 0) { return; }
+
+    unregister_VI_handler(joypad_vi_interrupt_callback);
+    
+    // Decrement the timer subsystem refcount (possibly closing it)
+    timer_close();
+}
+
+/**
+ * @brief Identify Joypads and wait for completion.
+ * 
+ * @param reset Whether to reset the devices.
+ */
+void joypad_identify_sync(bool reset)
+{
+    ASSERT_JOYPAD_INITIALIZED();
+    // Wait for pending identify/reset operation to resolve
+    while (joypad_identify_pending) { /* Spinlock */ }
+    // Enqueue this identify/reset operation
+    joypad_identify_async(reset);
+    // Wait for the operation to finish
+    while (joypad_identify_pending) { /* Spinlock */ }
+}
+
+/**
+ * @brief Read Joypad inputs and wait for completion.
+ * 
+ * Implicitly scans the read inputs to save you a step.
+ */
+void joypad_read_sync(void)
+{
+    ASSERT_JOYPAD_INITIALIZED();
+    joypad_read_async();
+    while (joypad_read_pending) { /* Spinlock */ }
+    joypad_scan();
+}
+
+/**
+ * @brief Fetch the current Joypad input state.
+ * 
+ * This function must be called once per frame, or any time after the
+ * Joypads have been read. After calling this function, you can read
+ * the Joypad state using the following functions:
+ * 
+ * * #joypad_get_inputs
+ * * #joypad_get_buttons
+ * * #joypad_get_buttons_pressed
+ * * #joypad_get_buttons_released
+ * * #joypad_get_buttons_held
+ * * #joypad_get_dpad_direction
+ * * #joypad_get_c_direction
+ * * #joypad_get_stick_direction
+ * * #joypad_get_axis_pressed
+ * * #joypad_get_axis_released
+ * * #joypad_get_axis_held
+ * 
+ * This function is very fast. In fact, joypads are read in background
+ * asynchronously under interrupt, so this function just synchronizes the
+ * internal state.
+ */
+void joypad_scan(void)
+{
+    ASSERT_JOYPAD_INITIALIZED();
+    // Bail early if the joypads have not been read since last call
+    static uint64_t prev_read_count = 0;
+    uint64_t read_count = joypad_read_count;
+    if (prev_read_count == read_count) { return; }
```

Comment:
I'm a bit concerned about this. If I understand correclty, this means that the pressed/held status of buttons will not be affected, and will keep returning the previous one. I think this breaks my mental model for joypad_scan(). In my model, pressed/held state always refers "since the previous joypad_scan()", but with this bailout the invariant seems broken.

Reply:
If I recall correctly, this was a defensive measure in case a developer called joypad_scan multiple times in a frame. I didn't see the point in re-scanning the inputs if they hadn't actually changed.

I'm not married to this check: it's an edge-case where the developer is most likely doing something wrong. I figured doing nothing was better than asserting, but perhaps the right answer is to let the developer do it wrong: it's basically harmless.

Reply:
I don't think calling joypad_scan multiple times is "wrong". As you say it's just harmless: as long as the edge-triggered event semantic is preserved, you get a fully correct representation of joypad inputs, which yes, haven't changed since last scan, but that's also true across frames as well, as the user's hands don't change inputs 60 times per second anyway.

Also in the future we might want to expose a developer knob to allow polling joypads more frequently (could be exposed as "N times per frame" for instance, with N=1 being the default). In that case, it might even make sense to call joypad_scan() frequently, without necessarily being able to prove whether a joybus transaction happened since last time or not (imagine the developer playing with the N above to try different settings -- during those tests, the condition "has a joybus transaction happened since last joypad_scan" can change between true and false semi-randomically).

Anyway, preserving the semantic means dropping this check, so I'd just remove it. With this check, none of the above is true, and whether a transaction has actually been made or not affects there results of pressed/released functions for instance, which is surely not what the user expects.

So I think I'd simply remove the check, that's probably the best coarse of action.

Reply:
Fixed in c2d0d6d3dc41f0489b0e2efe92a0136a415385b1